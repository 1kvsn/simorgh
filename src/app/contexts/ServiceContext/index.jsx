import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import variantPropType from '../../models/propTypes/variants';
import { getVariant } from '#lib/utilities/variantHandler';
import getLangOverride from '#lib/utilities/langHandler';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({
  children,
  service,
  variant,
  pageLang,
}) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getLangOverride(pageLang) || getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
  pageLang: string,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
  pageLang: null,
};
