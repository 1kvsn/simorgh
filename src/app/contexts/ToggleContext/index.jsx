import React, { createContext, useReducer } from 'react';
import { node, object } from 'prop-types';
import { toggleReducer } from './reducer';

const ToggleContext = createContext({});

const ToggleContextProvider = ({ children, toggles }) => {
  const [toggleState, toggleDispatch] = useReducer(toggleReducer, toggles);

  return (
    <ToggleContext.Provider value={{ toggleState, toggleDispatch }}>
      {children}
    </ToggleContext.Provider>
  );
};

const ToggleContextConsumer = ToggleContext.Consumer;

ToggleContextProvider.propTypes = {
  children: node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  toggles: object.isRequired,
};

export { ToggleContext, ToggleContextProvider, ToggleContextConsumer };
