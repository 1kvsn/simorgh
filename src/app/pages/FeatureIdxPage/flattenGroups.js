export default groups => {
  const numberOfGroups = groups.length;

  if (numberOfGroups < 2) {
    return groups;
  }

  const flattenedGroups = [];

  // eslint-disable-next-line no-plusplus
  for (let i = numberOfGroups - 1; i > 0; i--) {
    const group = groups[i];

    if (group.strapline) {
      flattenedGroups.push(group);
    } else {
      const previousGroup = groups[i - 1];
      previousGroup.items = [...previousGroup.items, ...group.items];
    }
  }

  flattenedGroups.push(groups[0]);

  return flattenedGroups.reverse();
};
