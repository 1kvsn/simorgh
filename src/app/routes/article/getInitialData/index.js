import pipe from 'ramda/src/pipe';
import identity from 'ramda/src/identity';

import isLive from '#lib/utilities/isLive';
import fetchPageData from '../../utils/fetchPageData';
import handleGroupBlocks from '../handleGroupBlocks';
import {
  augmentWithTimestamp,
  addIdsToBlocks,
  applyBlockPositioning,
  addIndexesToEmbeds,
} from '../../utils/sharedDataTransformers';

import getErrorStatusCode from '../../utils/fetchPageData/utils/getErrorStatusCode';

const transformJson = pipe(
  isLive() ? identity : handleGroupBlocks,
  augmentWithTimestamp,
  addIdsToBlocks,
  applyBlockPositioning,
  addIndexesToEmbeds,
);

export default async ({ path, pageType }) => {
  try {
    const { json, status } = await fetchPageData({
      path,
      pageType,
    });

    return {
      status,
      pageData: transformJson(json),
    };
  } catch ({ message, status = getErrorStatusCode() }) {
    return { error: message, status };
  }
};
