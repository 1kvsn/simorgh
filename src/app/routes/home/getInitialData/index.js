import pipe from 'ramda/src/pipe';
import pathOr from 'ramda/src/pathOr';
import fetchPageData from '#app/routes/utils/fetchPageData';
import withRadioSchedule from '#app/routes/utils/withRadioSchedule';
import getConfig from '#app/routes/utils/getConfig';
import filterUnknownContentTypes from '#app/routes/utils/sharedDataTransformers/filterUnknownContentTypes';
import filterEmptyGroupItems from '#app/routes/utils/sharedDataTransformers/filterEmptyGroupItems';
import squashTopStories from '#app/routes/utils/sharedDataTransformers/squashTopStories';
import addIdsToItems from '#app/routes/utils/sharedDataTransformers/addIdsToItems';
import filterGroupsWithoutStraplines from '#app/routes/utils/sharedDataTransformers/filterGroupsWithoutStraplines';
import getErrorStatusCode from '../../utils/fetchPageData/utils/getErrorStatusCode';

const transformJson = pipe(
  filterUnknownContentTypes,
  filterEmptyGroupItems,
  addIdsToItems,
  squashTopStories,
  filterGroupsWithoutStraplines,
);

export const hasRadioSchedule = async (service, variant) => {
  const config = await getConfig(service, variant);

  const serviceHasRadioSchedule = pathOr(
    false,
    ['radioSchedule', 'hasRadioSchedule'],
    config,
  );

  const radioScheduleOnFrontPage = pathOr(
    false,
    ['radioSchedule', 'onFrontPage'],
    config,
  );

  return serviceHasRadioSchedule && radioScheduleOnFrontPage;
};

const fetchElectionsOembed = async service => {
  const usElectionOembedPath = `/${service}/election/us2020/results/oembed`;

  try {
    const { json, status } = await fetchPageData({
      path: usElectionOembedPath,
    });

    if (json && status === 200) {
      return { usElectionOembed: json };
    }
  } catch (error) {
    return {};
  }

  return null;
};

export default async ({ path, service, variant, pageType, toggles }) => {
  try {
    const pageHasRadioSchedule = await hasRadioSchedule(service, variant);
    const pageDataPromise = fetchPageData({ path, pageType });
    const pageHasUsElectionsBanner = pathOr(
      false,
      ['us2020ElectionBanner'],
      toggles,
    );

    const { json, status } = pageHasRadioSchedule
      ? await withRadioSchedule({
          pageDataPromise,
          service,
          path,
          pageType: 'Home',
        })
      : await pageDataPromise;

    return {
      status,
      pageData: {
        ...transformJson(json),
        ...(pageHasUsElectionsBanner
          ? await fetchElectionsOembed(service)
          : null),
      },
    };
  } catch ({ message, status = getErrorStatusCode() }) {
    return { error: message, status };
  }
};
