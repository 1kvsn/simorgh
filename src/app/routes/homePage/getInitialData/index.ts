/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Agent } from 'https';
import Url from 'url-parse';
import nodeLogger from '../../../lib/logger.node';
import { BFF_FETCH_ERROR } from '../../../lib/logger.const';
import fetchPageData from '../../utils/fetchPageData';
import { Services } from '../../../models/types/global';
import HOME_PAGE_CONFIG from './page-config';

const logger = nodeLogger(__filename);

const getEnvironment = (pathname: string) => {
  if (pathname.includes('renderer_env=test')) {
    return 'test';
  }
  if (pathname.includes('renderer_env=live')) {
    return 'live';
  }

  return process.env.SIMORGH_APP_ENV;
};

interface BFFError extends Error {
  status: number;
}

const handleError = (message: string, status: number) => {
  const error = new Error(message) as BFFError;
  error.status = status;

  return error;
};

type Props = {
  getAgent: () => Promise<Agent>;
  service: Services;
  path: string;
  pageType: 'homePage';
};

export default async ({ getAgent, service, path: pathname }: Props) => {
  try {
    const env = getEnvironment(pathname);
    const isLocal = !env || env === 'local';
    const idsForService = HOME_PAGE_CONFIG[service];

    const agent = !isLocal ? await getAgent() : null;
    const id = env === 'live' ? idsForService.live : idsForService.test;

    if (!id) throw handleError('Home ID is invalid', 500);

    let fetchUrl = Url(process.env.BFF_PATH as string).set('query', {
      id,
      service,
      pageType: 'home',
    });

    const optHeaders = { 'ctx-service-env': env };

    if (isLocal) {
      fetchUrl = Url(`/${service}/tipohome`);
    }

    // @ts-ignore - Ignore fetchPageData argument types
    // eslint-disable-next-line prefer-const
    let { status, json } = await fetchPageData({
      path: fetchUrl.toString(),
      ...(!isLocal && { agent, optHeaders }),
    });

    const { data } = json;

    return {
      status,
      pageData: {
        id,
        title: data.title,
        pageType: 'home',
        curations: data.curations,
      },
    };
  } catch ({ message, status }) {
    logger.error(BFF_FETCH_ERROR, {
      service,
      status,
      pathname,
      message,
    });
    return { error: message, status };
  }
};
