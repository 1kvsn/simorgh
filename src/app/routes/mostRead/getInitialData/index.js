import fetchPageData from '../../utils/fetchPageData';
import { getMostReadEndpoint } from '#lib/utilities/getMostReadUrls';

export default async ({ service, variant }) => {
  try {
    const mostReadUrl = getMostReadEndpoint({ service, variant }).split('.')[0];
    const { json, status } = await fetchPageData(mostReadUrl);
    const pageTypeMeta = { metadata: { type: 'mostRead' } };

    return {
      status,
      pageData: { ...json, ...pageTypeMeta },
    };
  } catch ({ error = true, status = 500 }) {
    return { error, status };
  }
};
