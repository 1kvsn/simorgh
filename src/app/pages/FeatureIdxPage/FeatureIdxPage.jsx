import React, { useContext, Fragment } from 'react';
import path from 'ramda/src/path';
import styled from 'styled-components';
import { node, string } from 'prop-types';
import { GEL_GROUP_4_SCREEN_WIDTH_MIN } from '@bbc/gel-foundations/breakpoints';
import { ServiceContext } from '#contexts/ServiceContext';
import MetadataContainer from '#containers/Metadata';
import LinkedData from '#containers/LinkedData';
import IndexHeading from '#containers/IndexHeading';
import IndexPageContainer from '#app/components/PageLayout/IndexPageContainer';
import MostReadContainer from '#containers/MostRead';
import MostReadSection from '#containers/MostRead/section';
import MostReadSectionLabel from '#containers/MostRead/label';
import IndexPageSection from '#containers/IndexPageSection';
import idxPageDataPropTypes from '#models/propTypes/idxPage';
import ATIAnalytics from '#containers/ATIAnalytics';
import ChartbeatAnalytics from '#containers/ChartbeatAnalytics';
import ComscoreAnalytics from '#containers/ComscoreAnalytics';

const FeatureIdxMostReadSection = styled(MostReadSection)`
  /* To centre page layout for Group 4+ */
  margin: 0 auto;
  width: 100%; /* Needed for IE11 */
  @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) {
    max-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN};
  }
`;

const MostReadWrapper = ({ children }) => (
  <FeatureIdxMostReadSection>
    <MostReadSectionLabel />
    {children}
  </FeatureIdxMostReadSection>
);

const renderMostRead = mostReadEndpointOverride => (
  <MostReadContainer
    mostReadEndpointOverride={mostReadEndpointOverride}
    columnLayout="twoColumn"
    wrapper={MostReadWrapper}
  />
);

MostReadWrapper.propTypes = {
  children: node.isRequired,
};

const FeatureIdxPage = ({ pageData, mostReadEndpointOverride }) => {
  const { mostRead, lang } = useContext(ServiceContext);

  const groups = path(['content', 'groups'], pageData);
  const title = path(['metadata', 'title'], pageData);
  const summary = path(['metadata', 'summary'], pageData);
  const seoTitle = path(['promo', 'name'], pageData);
  const mostReadOnIdxPage = path(['onIdxPage'], mostRead);

  return (
    <>
      <ATIAnalytics data={pageData} />
      <ChartbeatAnalytics data={pageData} />
      <ComscoreAnalytics />
      <MetadataContainer
        title={title}
        lang={lang}
        description={summary}
        openGraphType="website"
      />
      <LinkedData type="WebPage" seoTitle={seoTitle} />
      <main role="main">
        <IndexPageContainer>
          <IndexHeading id="content" pageType="idx">
            {title}
          </IndexHeading>
          {groups.map((group, index) => (
            <Fragment key={group.title}>
              <IndexPageSection group={group} sectionNumber={index} />
            </Fragment>
          ))}
          {mostReadOnIdxPage && renderMostRead(mostReadEndpointOverride)}
        </IndexPageContainer>
      </main>
    </>
  );
};

FeatureIdxPage.propTypes = {
  pageData: idxPageDataPropTypes.isRequired,
  mostReadEndpointOverride: string,
};

FeatureIdxPage.defaultProps = {
  mostReadEndpointOverride: null,
};

export default FeatureIdxPage;
