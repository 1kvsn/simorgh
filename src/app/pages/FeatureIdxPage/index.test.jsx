/* eslint-disable react/prop-types */
import React from 'react';
import fetchMock from 'fetch-mock';
import { BrowserRouter } from 'react-router-dom';
import { render, act } from '@testing-library/react';
import assocPath from 'ramda/src/assocPath';
import { RequestContextProvider } from '#contexts/RequestContext';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { ToggleContextProvider } from '#contexts/ToggleContext';
import afriqueFeatureIdxPageData from '#data/afrique/cpsAssets/48465371';
import getInitialData from '#app/routes/cpsAsset/getInitialData';
import FeatureIdxPage from '.';

const mockToggles = {
  comscoreAnalytics: {
    enabled: true,
  },
  ads: {
    enabled: false,
  },
};

const requestContextData = ({
  service = 'afrique',
  showAdsBasedOnLocation,
}) => ({
  pageType: 'FIX',
  service,
  pathname: '/pathname',
  data: { status: 200 },
  showAdsBasedOnLocation,
});

// eslint-disable-next-line react/prop-types
const FeatureIdxPageWithContext = ({
  isAmp = false,
  service = 'afrique',
  toggles = mockToggles,
  showAdsBasedOnLocation,
  ...props
}) => (
  <BrowserRouter>
    <ToggleContextProvider toggles={toggles}>
      <RequestContextProvider
        isAmp={isAmp}
        {...requestContextData({ service, showAdsBasedOnLocation })}
      >
        <ServiceContextProvider service={service}>
          <FeatureIdxPage {...props} />
        </ServiceContextProvider>
      </RequestContextProvider>
    </ToggleContextProvider>
  </BrowserRouter>
);

jest.mock('uuid', () => {
  let x = 1;
  return () => {
    x += 1;
    return `mockid-${x}`;
  };
});

jest.mock('#containers/ChartbeatAnalytics', () => {
  return () => <div>chartbeat</div>;
});

jest.mock('#containers/ATIAnalytics/amp', () => {
  return () => <div>Amp ATI analytics</div>;
});

jest.mock('#containers/PageHandlers/withVariant', () => Component => {
  return props => (
    <div id="VariantContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withContexts', () => Component => {
  return props => (
    <div id="ContextsContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withPageWrapper', () => Component => {
  return props => (
    <div id="PageWrapperContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withLoading', () => Component => {
  return props => (
    <div id="LoadingContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withError', () => Component => {
  return props => (
    <div id="ErrorContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withData', () => Component => {
  return props => (
    <div id="DataContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/PageHandlers/withContexts', () => Component => {
  return props => (
    <div id="ContextsContainer">
      <Component {...props} />
    </div>
  );
});

jest.mock('#containers/Ad/MPU', () => {
  const MPUContainer = () => (
    <div data-testid="fix-ads">Feature Index Page Ad - MPU</div>
  );
  return MPUContainer;
});

jest.mock('#containers/Ad/Canonical/CanonicalAdBootstrapJs', () => {
  const CanonicalAdBootstrapJs = () => (
    <div data-testid="adBootstrap">bootstrap</div>
  );
  return CanonicalAdBootstrapJs;
});

describe('Feature Idx Page', () => {
  let pageData;

  beforeEach(async () => {
    fetchMock.mock(
      'http://localhost/some-feature-idx-page-path.json',
      JSON.stringify(afriqueFeatureIdxPageData),
    );

    ({ pageData } = await getInitialData({
      path: 'some-feature-idx-page-path',
      service: 'afrique',
    }));
  });

  afterEach(() => {
    fetchMock.restore();
    jest.clearAllMocks();
  });

  describe('snapshots', () => {
    it('should render an afrique feature idx page correctly', async () => {
      let container;
      await act(async () => {
        container = render(<FeatureIdxPageWithContext pageData={pageData} />)
          .container;
      });
      expect(container).toMatchSnapshot();
    });

    it('should render an afrique amp feature idx page', async () => {
      let container;
      await act(async () => {
        container = render(
          <FeatureIdxPageWithContext pageData={pageData} isAmp />,
        ).container;
      });
      expect(container).toMatchSnapshot();
    });
  });

  describe('Assertions', () => {
    it('should render visually hidden text as h1', async () => {
      let container;
      await act(async () => {
        container = render(<FeatureIdxPageWithContext pageData={pageData} />)
          .container;
      });

      const h1 = container.querySelector('h1');
      const content = h1.getAttribute('id');
      const tabIndex = h1.getAttribute('tabIndex');

      expect(content).toEqual('content');
      expect(tabIndex).toBe('-1');

      // const span = h1.querySelector('span');
      // expect(span.getAttribute('role')).toEqual('text');
      expect(h1.textContent).toEqual('Tout savoir sur la CAN 2019');

      // const langSpan = span.querySelector('span');
      // expect(langSpan.getAttribute('lang')).toEqual('en-GB');
      // expect(langSpan.textContent).toEqual('BBC News');
    });

    it('should render feature index page sections', async () => {
      let container;
      await act(async () => {
        container = render(<FeatureIdxPageWithContext pageData={pageData} />)
          .container;
      });

      const sections = container.querySelectorAll('section');
      expect(sections).toHaveLength(10);
      sections.forEach(section => {
        expect(section.getAttribute('role')).toEqual('region');
      });
    });

    describe('Ads', () => {
      it.each`
        adsEnabled | showAdsBasedOnLocation | scenario
        ${false}   | ${true}                | ${'ads feature toggle is not enabled'}
        ${true}    | ${false}               | ${'ads not permitted to be shown in location'}
      `(
        'should not render because $scenario',
        async ({ showAdsBasedOnLocation, adsEnabled }) => {
          const toggles = {
            ads: {
              enabled: adsEnabled,
            },
          };

          let queryByTestId;
          await act(async () => {
            ({ queryByTestId } = render(
              <FeatureIdxPageWithContext
                pageData={pageData}
                showAdsBasedOnLocation={showAdsBasedOnLocation}
                toggles={toggles}
              />,
            ));
          });

          const fixPageAds = queryByTestId('fix-ads');
          expect(fixPageAds).not.toBeInTheDocument();
          const adBootstrap = queryByTestId('adBootstrap');
          expect(adBootstrap).not.toBeInTheDocument();
        },
      );

      it('should render leaderboard and MPU ads when the ads toggle is enabled and page has top-stories', async () => {
        const toggles = {
          ads: {
            enabled: true,
          },
        };

        const pageDataWithTopStories = assocPath(
          ['content', 'groups', 0, 'type'],
          'top-stories',
          pageData,
        );

        let getByTestId;
        await act(async () => {
          ({ getByTestId } = render(
            <FeatureIdxPageWithContext
              pageData={pageDataWithTopStories}
              showAdsBasedOnLocation
              toggles={toggles}
            />,
          ));
        });

        const leaderboardAd = document.querySelector(
          '[data-e2e="advertisement"]',
        );
        expect(leaderboardAd).toBeInTheDocument();

        const mpuAd = getByTestId('fix-ads');
        expect(mpuAd).toBeInTheDocument();

        const adBootstrap = getByTestId('adBootstrap');
        expect(adBootstrap).toBeInTheDocument();
      });

      it('should not render canonical ad bootstrap on amp', async () => {
        process.env.SIMORGH_APP_ENV = 'test';
        const toggles = {
          ads: {
            enabled: true,
          },
        };

        let queryByTestId;
        await act(async () => {
          ({ queryByTestId } = render(
            <FeatureIdxPageWithContext
              pageData={pageData}
              toggles={toggles}
              showAdsBasedOnLocation={false}
            />,
          ));
        });

        const adBootstrap = queryByTestId('adBootstrap');
        expect(adBootstrap).not.toBeInTheDocument();
      });
    });
  });
});
