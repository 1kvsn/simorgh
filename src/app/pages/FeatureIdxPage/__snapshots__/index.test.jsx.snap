// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature Idx Page snapshots should render a afrique feature idx page correctly 1`] = `
.c1 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-family: Helmet,Freesans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  color: #6E6E73;
  margin: 0;
  padding-bottom: 1.5rem;
}

.c0 {
  margin: 0 0.5rem;
  padding-top: 0.5rem;
  padding-bottom: 2rem;
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c1 {
    font-size: 1.25rem;
    line-height: 1.625rem;
  }
}

@media (min-width:37.5rem) {
  .c1 {
    font-size: 1.5rem;
    line-height: 2rem;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c1 {
    padding: 1.5rem 0 0.5rem;
  }
}

@media (min-width:63rem) {
  .c1 {
    padding: 1.5rem 0 0;
  }
}

@media (min-width:80rem) {
  .c1 {
    width: 100%;
    margin: 0 auto;
    max-width: 80rem;
  }
}

@media (min-width:25rem) {
  .c0 {
    margin: 0 1rem;
  }
}

@media (min-width:25rem) {
  .c0 {
    padding-top: 1rem;
  }
}

@media (max-width:37.4375rem) {
  .c0 {
    padding-bottom: 1.5rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    padding-top: 0;
  }
}

@media (min-width:63rem) {
  .c0 {
    padding-bottom: 2.5rem;
  }
}

<div>
  <div
    id="ContextsContainer"
  >
    <div
      id="PageWrapperContainer"
    >
      <div
        id="LoadingContainer"
      >
        <div
          id="ErrorContainer"
        >
          <div
            id="DataContainer"
          >
            <div>
              chartbeat
            </div>
            <main
              role="main"
            >
              <div
                class="c0"
              >
                <h1
                  class="c1"
                  dir="ltr"
                  id="content"
                  tabindex="-1"
                >
                  Tout savoir sur la CAN 2019
                </h1>
              </div>
            </main>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
